crates/noirc_evaluator/src/ssa/optim.rs

```diff
@@ -382,53 +382,74 @@ pub fn block_cse(
             anchor_push(ins.operator, anchor);
             if node::is_binary(ins.operator) {
                 //binary operation:
-                i_lhs = propagate(ctx, ins.lhs);
-                i_rhs = propagate(ctx, ins.rhs);
+                i_lhs = propagate(ctx, ins.lhs, modified);
+                i_rhs = propagate(ctx, ins.rhs, modified);
                 if let Some(j) = find_similar_instruction(ctx, i_lhs, i_rhs, &anchor[&ins.operator])
                 {
                     to_delete = true; //we want to delete ins but ins is immutable so we use the new_list instead
                     i_rhs = j;
+                    debug_assert!(j != ins.id);
+                    *modified = true;
                 } else {
                     new_list.push(*iter);
                     anchor.get_mut(&ins.operator).unwrap().push_front(*iter);
                 }
             } else {
                 match ins.operator {
-                    node::Operation::Load(_) | node::Operation::Store(_) => {
-                        i_lhs = propagate(ctx, ins.lhs);
-                        i_rhs = propagate(ctx, ins.rhs);
+                    node::Operation::Load(x) | node::Operation::Store(x) => {
+                        if !is_join && ins.is_dummy() && ins.operator == node::Operation::Store(x) {
+                            continue;
+                        }
+                        i_lhs = propagate(ctx, ins.lhs, modified);
+                        i_rhs = propagate(ctx, ins.rhs, modified);
+
+                        let prev_ins = &anchor[&node::Operation::Load(x)];
+
                         let (cse_id, cse_action) = find_similar_mem_instruction(
                             ctx,
                             ins.operator,
                             ins.id,
                             i_lhs,
                             i_rhs,
-                            anchor,
+                            prev_ins,
                         );
+
                         match cse_action {
-                            CseAction::Keep => new_list.push(*iter),
+                            CseAction::Keep => {
+                                anchor
+                                    .get_mut(&node::Operation::Load(x))
+                                    .unwrap()
+                                    .push_front(*iter);
+                                new_list.push(*iter);
+                            }
                             CseAction::Replace => {
                                 to_delete = true;
                                 i_rhs = cse_id;
+                                *modified = true;
                             }
                             CseAction::Remove => {
+                                anchor
+                                    .get_mut(&node::Operation::Load(x))
+                                    .unwrap()
+                                    .push_front(*iter);
```

Comment:
`anchor[&Operation::Load(x)].push_front(*iter)`

Reply:
With your suggestion, the compiler complains that it `cannot borrow as mutable`.

Reply:
Ah, whoops my bad. Forgot `IndexMut` is not implemented for HashMaps.